src/debugAdapter/goDebug.ts

```diff
@@ -230,15 +230,20 @@ class Delve {
 
 			let env = Object.assign({}, process.env, fileEnv, launchArgsEnv);
 
+			let dirname = isProgramDirectory ? program : path.dirname(program);
 			if (!fileEnv['GOPATH'] && !launchArgsEnv['GOPATH'] && (mode === 'debug' || mode === 'test')) {
 				// If user hasnt specified GOPATH & file/package to debug is not under env['GOPATH'], then infer it from the file/package path
 				// Not applicable to exec mode in which case `program` need not point to source code under GOPATH
-				let programNotUnderGopath = !env['GOPATH'] || !getCurrentGoWorkspaceFromGOPATH(env['GOPATH'], isProgramDirectory ? program : path.dirname(program));
+				let currentGOWorkspace = getCurrentGoWorkspaceFromGOPATH(env['GOPATH'], dirname);
+				let programNotUnderGopath = !env['GOPATH'] || !currentGOWorkspace;
 				if (programNotUnderGopath ) {
-					env['GOPATH'] = getInferredGopath(isProgramDirectory ? program : path.dirname(program)) || env['GOPATH'];
+					env['GOPATH'] = getInferredGopath(dirname) || env['GOPATH'];
 				}
 			}
 
+			let currentGOWorkspace = getCurrentGoWorkspaceFromGOPATH(env['GOPATH'], dirname);
+			let importPath = currentGOWorkspace ? dirname.substr(currentGOWorkspace.length + 1) : '.';
```

Comment:
shouldnt we do this only if `mode` is `debug` or `test`?
Other values for `mode` are `remote` and `exec`